Adding a Java autoscaler example.

sduskis opened the pull request.

waprin commented on java/autoscaler/pom.xml:
```diff
@@ -0,0 +1,87 @@
+    <dependency>
+      <groupId>com.google.apis</groupId>
+      <artifactId>google-api-services-monitoring</artifactId>
+      <version>v3-rev400-1.22.0</version>
```
This is not the recommended client library for Java Stackdriver Monitoring. [This one is](https://github.com/GoogleCloudPlatform/google-cloud-java/tree/master/google-cloud-monitoring), it's the new gRPC based autogenerated ones.  I don't think the google-api-java-client libs are going anywhere anytime soon, and the new library is still in alpha...but it's worth noting that this sample is using a library we are moving away from.

waprin commented on java/autoscaler/src/main/java/com/example/bigtable/autoscaler/Autoscaler.java:
```diff
@@ -0,0 +1,207 @@
+
+/**
+ * Copyright 2017 Google Inc. All Rights Reserved.
```
nit - we use single star for license headers since it's not javadoc

waprin commented on java/autoscaler/src/main/java/com/example/bigtable/autoscaler/Autoscaler.java:
```diff
@@ -0,0 +1,207 @@
+              clusterUtility.setClusterSize(clusterId, zoneId,
+                Math.max(clusterSize - 3, MIN_NODE_COUNT));
+            }
+          } else if (latestValue > .7) {
```
prob want to refactor out .7 and .5 to constants or configs?

hegemonic commented on java/autoscaler/src/main/java/com/example/bigtable/autoscaler/Autoscaler.java:
```diff
@@ -0,0 +1,207 @@
+
+  /**
+   * The minimum number of nodes to use. The default minimum is 3. If you have a lot of data, the
+   * rule of thumb is to not go below 1 node per terabyte for SSD clusters, and 5 terabytes for HDD.
```
FYI, Misha and Eric told me that the rule of thumb was 1 node per 2.5 TB for SSD and 1 node per 8 TB for HDD. It'd be good to confirm the numbers with them so this comment is consistent with our docs.

sduskis commented:
All the comments were addressed.  PTAL.  Next step, add a README.

waprin commented:
Code LGTM from my perspective if you want to merge. Test coverage would be good ,though unit tests aren't going to do that much and integration tests are flakey/difficult to write (as all sample code faces). Can try to help with that.

I will do some manual testing of this as well

waprin commented:
@sduskis working on finishing this, just an FYI so we don't step on each other's toes. I think biggest change is we need sleep after a cluster resize to allow changes to take effect (as is will start rapidly scaling up cluster without waiting). Also adding a few tests and readme and region tags. Thanks for starting point.

Btw, any objections to moving to java-docs-samples? Same as python, those repos get a lot more maintenance love so prob worth it.

waprin commented:
Created https://github.com/sduskis/cloud-bigtable-examples/pull/1 to cherry-pick into here.

As far as testing, talked to @lesv about what standards are, sounds like we'll skip unit tests but it would be good if we at least have a script in `travis.sh` that at least kicks off the autoscaler and makes sure it doesn't barf, a full system test is probably to complicated but at least that's something, so I'll look into that.

waprin commented:
Oh and also we need to consider whether we should changed Autoscaler to MetricScaler or something to make it clear this is programatic scaling rather than true automatic scaling.

sduskis commented:
I updated the name of the class and project to metric scaler.  I added the project to the top pom.xml so that the travis build will do a sanity check on the project.  I'm committing the change.
